### PR TITLE
feat: Add McNaughton–Yamada–Thompson algorithm

### PR DESCRIPTION
**Pull request by CascadingCascade**
#### Description of Change
Added a NFA regex matcher based on the [McNaughton–Yamada–Thompson algorithm](https://en.wikipedia.org/wiki/Thompson%27s_construction)

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
It's been awhile since my last PR, and this is what I've cooked up.:smile: I tried my best, but a program this large is bound to have more bugs than what I managed to found, thanks in advance for sifting through ~700 lines of code.

### CODE REVIEW

File: misc/mcnaughton_yamada_thompson.c [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+    testHelper("(0|(1(01*(00)*0)*1)*)*","1100",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","10000",0);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","00000",1);
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(void) {
```
Comment (Panquesito7) [inline misc/mcnaughton_yamada_thompson.c:591]: Unnecessary, IMO.
```suggestion
int main() {
```

Comment (CascadingCascade) [reply misc/mcnaughton_yamada_thompson.c:591]: Actually, I specifically research this some time ago, and it turns out there is a difference between `int main()` and `int main(void)`. `int main()` accepts an unspecified amount of arguments while `int main(void)` explicitly accepts no arguments. If you crank the warning level high enough, your compiler / IDE will complain about `function declaration without prototype is deprecated in all versions of C`. [Source](https://stackoverflow.com/questions/42125/warning-error-function-declaration-isnt-a-prototype)

Comment (CascadingCascade) [reply misc/mcnaughton_yamada_thompson.c:591]: BTW, I code with this much warning on: `add_compile_options(-Wall -Wextra -Wpedantic -Wformat=2 -Wno-unused-parameter
        -Wshadow -Wwrite-strings -Wstrict-prototypes -Wold-style-definition
        -Wredundant-decls -Wnested-externs -Wmissing-include-dirs -Wjump-misses-init -Wlogical-op)`

File: misc/mcnaughton_yamada_thompson.c [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+    testHelper("(c|a*b)", "ca", 0);
+    testHelper("(c|a*b)*", "caaab", 1);
+    testHelper("(c|a*b)*", "caba", 0);
+    testHelper("", "", 1);
+    testHelper("", "1", 0);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","11",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","110",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","1100",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","10000",0);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","00000",1);
```
Comment (Panquesito7) [inline misc/mcnaughton_yamada_thompson.c:584]:
```suggestion
    testHelper("(0|(1(01*(00)*0)*1)*)*","00000",1);

    printf("All tests have successfully passed!\n");
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work. Thanks! 🚀

**Review by alexpantyukhin — APPROVED** [review]
This is cool!

**Pull request merged**